framework/src/problems/FEProblemBase.C

```diff
@@ -4179,6 +4184,348 @@ FEProblemBase::solve()
+               << '\n';
+
+      if (picard_it ==
+          0) // First Picard iteration - need to save off the initial nonlinear residual
+      {
+        picard_initial_norm = computeResidualL2Norm();
+        _console << COLOR_MAGENTA << "Initial Picard Norm: " << COLOR_DEFAULT;
+        if (picard_initial_norm == std::numeric_limits<Real>::max())
+          _console << " MAX ";
+        else
+          _console << std::scientific << picard_initial_norm;
+        _console << COLOR_DEFAULT << '\n';
+      }
+      else
+      {
+        // For every iteration other than the first, we need to restore the state of the MultiApps
+        restoreMultiApps(EXEC_TIMESTEP_BEGIN);
+        restoreMultiApps(EXEC_TIMESTEP_END);
+      }
+    }
+
+    Real begin_norm_old = (picard_it > 0 ? picard_timestep_begin_norm[picard_it - 1]
+                                         : std::numeric_limits<Real>::max());
+    Real end_norm_old = (picard_it > 0 ? picard_timestep_end_norm[picard_it - 1]
+                                       : std::numeric_limits<Real>::max());
+    bool relax = (solver_params._picard_relaxation_factor != 1) && (picard_it > 0);
+    bool solve_converged = solveStep(begin_norm_old,
+                                     picard_timestep_begin_norm[picard_it],
+                                     end_norm_old,
+                                     picard_timestep_end_norm[picard_it],
+                                     relax,
+                                     relaxed_dofs);
+
+    if (solve_converged)
+    {
+      if (solver_params._has_picard_its)
+      {
+        _console << "\n 0 Picard |R| = "
+                 << Console::outputNorm(std::numeric_limits<Real>::max(), picard_initial_norm)
+                 << '\n';
+
+        for (unsigned int i = 1; i <= picard_it; ++i)
+        {
+          Real max_norm = std::max(picard_timestep_begin_norm[i], picard_timestep_end_norm[i]);
+          _console << std::setw(2) << i
+                   << " Picard |R| = " << Console::outputNorm(picard_initial_norm, max_norm)
+                   << '\n';
+        }
+
+        Real max_norm =
+            std::max(picard_timestep_begin_norm[picard_it], picard_timestep_end_norm[picard_it]);
+
+        Real max_relative_drop = max_norm / picard_initial_norm;
+
+        if (max_norm < solver_params._picard_abs_tol)
+        {
+          _picard_status = MOOSE_PICARD_CONVERGED_ABS;
+          break;
+        }
+        if (max_relative_drop < solver_params._picard_rel_tol)
+        {
+          _picard_status = MOOSE_PICARD_CONVERGED_RELATIVE;
+          break;
+        }
+        if (extraPicardConvergenceCheck())
+        {
+          _picard_status = MOOSE_PICARD_CONVERGED_CUSTOM;
+          break;
+        }
+        if (picard_it + 1 == solver_params._picard_max_its)
+        {
+          _picard_status = MOOSE_PICARD_DIVERGED_MAX_ITS;
+          converged = false;
+          break;
+        }
+      }
+    }
```

Comment:
This looks pretty good, however we could probably still continue to factor out more of this code into methods just so these loops and branch fit on a developer's screen when they are looking at this code. You don't necessarily need to do that in this PR. There's a lot of refactoring going on already.